torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,196 @@
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* How can we make these ops work in eager without manually enabling python_dispatacher mode?
+
+"""
+
+"""
+FIXME for this to work correctly we need to change Work to internally hold no reference to the tensor.
+"""
+
+tensor_to_work = WeakIdKeyDictionary()
+
+lib = torch.library.Library("tr_c10d", "DEF")
+lib.define("wait(Tensor self) -> Tensor")
+
+impl_lib_cpu = torch.library.Library("tr_c10d", "IMPL", "CPU")
+impl_lib_cuda = torch.library.Library("tr_c10d", "IMPL", "CUDA")
+
+def _wait_tensor(tensor: torch.Tensor):
+    w = tensor_to_work.get(tensor)
+    if w:
+        del tensor_to_work[tensor]
+        w.wait()
+    return tensor
+
+impl_lib_cpu.impl("wait", _wait_tensor)
+impl_lib_cuda.impl("wait", _wait_tensor)
+
+def wait_tensor(tensor):
+    return torch._ops.ops.tr_c10d.wait(tensor)
+
+class AsyncCollectiveTensor(torch.Tensor):
+    r"""
+    A Tensor subclass that is only used in eager mode, to hold a 'work' object
+    and then wait on it before invoking a real op.
+
+    Usage, from inside functional collective:
+    def functional_collective(input):
+        input = input.clone()
+        mutated_input, work = c10d.{inplace_collective}(input)
+        return AsyncCollectiveTensor(mutated_input, work)
+    """
+    _tensor: torch.Tensor
+
+    # disable __torch_function__ so that CommTensor can recursively dispatch
+    # with ProxyTorchDispatchMode in make_fx
+    __torch_function__ = _disabled_torch_function_impl
+
+    @staticmethod
+    def __new__(cls, tensor: torch.Tensor):
+        t = tensor
+        r = torch.Tensor._make_subclass(cls, t, require_grad=t.requires_grad)
+        r._tensor = tensor  # type: ignore[attr-defined]
+        return r
+
+    def __repr__(self):
+        return f"AsyncCollectiveTensor({self._tensor})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap(e: Any):
+            if isinstance(e, AsyncCollectiveTensor):
+                return wait_tensor(e._tensor)
+            return e
+
+        unwrapped_args = tree_map(unwrap, args)
+        unwrapped_kwargs = tree_map(unwrap, kwargs)
+
+        out = func(*unwrapped_args, **unwrapped_kwargs)
+        return out
+
+def _str_to_reduce_op(reduceOp: str) -> dist.ReduceOp:
+    reduceOp = reduceOp.upper()
+    op = dist.ReduceOp.RedOpType.__members__.get(reduceOp)
+    if op is None:
+        raise ValueError(f"Invalid reduce operation {reduceOp}")
+    return cast(dist.ReduceOp, op)
+
+# TODO assert if ranks has duplicated entries
+def _all_reduce(self, reduceOp, tag, ranks, stride):
+    op = _str_to_reduce_op(reduceOp)
+    group = c10d._find_or_create_pg_by_ranks_and_tag(tag, ranks, stride)
+    assert group is not None
+
+    inplace_tensor = self.clone()
```

Comment:
I am trying to understand when does it make sense to use this in eager? Looks like this will always incur additional overhead in memory access and footprint. 

@wanchaol @kumpera would I be correct that DTensor will use this so that we can directly trace DTensor collectives into the graph instead of inserting nodes? If we do that, will this jeopardize DTensor's perf?

Reply:
But this will still be tricky for things like allgather, where inputs and outputs are not the same. In that case, if outputs are not passed in, we will have to create output tensors within this functional call and the eager perf is at CUDACachingAllocator's mercy.

Reply:
> But this will still be tricky for things like allgather, where inputs and outputs are not the same. I

This should be fine, we can do tracing detection in DTensor.

Reply:
Yes DTensor would use functional collective as it's already doing the collective in a functional way. I think we would have to discuss how all_gather, reduce_scatter, and alltoall signature of the functional collective should be as they are a bit tricky, especially if it's uneven.

Reply:
> The downside of this design is that users must use the result tensor regardless to ensure their code works correctly.

for all_reduce, would you propose `result = all_reduce(input, inplace=True)`

or ``result = all_reduce(input, inplace=True, out=input)`

it seems confusing to have both.

But for an op like allgather, i guess we would offer `out` _instead of_ `inplace=True`?

Reply:
I meant that we would offer one of the other. It might make more sense to provide just `out` as it handles all collectives unlike `inplace=True`

Reply:
i see. hmm. i think technically we can offer just `out` for all collectives.  but for ops like allreduce it means you could pass a different input and output buffer and we should have an opinion what to do in that case.